sound/soc/sof/sof-audio.c

```diff
@@ -43,12 +43,43 @@ static int sof_kcontrol_setup(struct snd_sof_dev *sdev, struct snd_sof_control *
 	return ret;
 }
 
-static int sof_dai_config_setup(struct snd_sof_dev *sdev, struct snd_sof_dai *dai)
+static struct sof_ipc_dai_config* sof_dai_config_match(struct snd_sof_dev *sdev,
+						       struct snd_sof_dai *dai,
+						       struct sof_ipc_pcm_params *ipc_pcm)
 {
+	int sample_rate;
+	int num_conf;
+	int i;
+
+	if (!ipc_pcm)
+		return &dai->dai_config[dai->def_conf];
+
+	sample_rate = ipc_pcm->params.rate;
+	num_conf = dai->num_conf;
+
+	/* match for ssp sample rate and fsync for now, otherwise use default config */
+	for (i = 0; i < num_conf; i++) {
+		if (dai->dai_config[i].type == SOF_DAI_INTEL_SSP &&
+		    sample_rate == dai->dai_config[i].ssp.fsync_rate) {
+			dev_dbg(sdev->dev, "found dai config match at index %d\n", i);
```

Comment:
Humm, what happens if there is a sample-rate converted in the firmware? The sample rate is what you get for your PCM device, it may or may not be related directly to the SSP configuration.

The example would be BT HFP always working at 16kHz, with a 2x upsampling if we have 8kHz speech.

Something's not right here. 

I think you would need to let the firmware provide the response of which configuration it selected instead of making assumptions in the kernel. Then the kernel can propagate the information to the codec, etc.

Reply:
@plbossart the whole point of this PR is that we have a chip on the other end of SSP that changes the SSP configuration depending on the use case.

Firmware doesn't "know" which configuration to choose, all information is at user space.

- Ok, so we need to parse the multiple configs, as we can't sent them easily after the boot/topology loading.
- maybe we could have some enum control etc which would choose the default control -> and default control would be always send to dsp -> no matching in kernel?